### PR TITLE
KAFKA-14462; [1/N] Add new server configurations (KIP-848)

### PR DESCRIPTION
**Pull request by dajac**
This patch adds the new server configurations introduced in KIP-848. All of them are kept as internal configurations for now. We will make them public when the KIP is ready. It also adds an internal configuration named `group.coordinator.new.enable` that we will used during the development of the KIP. This one will be remove later on and replaced by IBP/feature flag.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 1]
```diff
@@ -1267,6 +1314,24 @@ object KafkaConfig {
...
+      // This property meant to be here only during the development of KIP-848. It will
+      // be replaced by a metadata version before releasing it.
+      .defineInternal(NewGroupCoordinatorEnableProp, BOOLEAN, Defaults.NewGroupCoordinatorEnable, null, MEDIUM, NewGroupCoordinatorEnableDoc)
+      .defineInternal(GroupCoordinatorNumThreadsProp, INT, Defaults.GroupCoordinatorNumThreads, atLeast(1), MEDIUM, GroupCoordinatorNumThreadsDoc)
+
+      /** Consumer groups configs */
+      // All properties are kept internal until KIP-848 is released.
+      .defineInternal(ConsumerGroupSessionTimeoutMsProp, INT, Defaults.ConsumerGroupSessionTimeoutMs, atLeast(1), MEDIUM, ConsumerGroupSessionTimeoutMsDoc)
+      .defineInternal(ConsumerGroupMinSessionTimeoutMsProp, INT, Defaults.ConsumerGroupMinSessionTimeoutMs, atLeast(1), MEDIUM, ConsumerGroupMinSessionTimeoutMsDoc)
+      .defineInternal(ConsumerGroupMaxSessionTimeoutMsProp, INT, Defaults.ConsumerGroupMaxSessionTimeoutMs, atLeast(1), MEDIUM, ConsumerGroupMaxSessionTimeoutMsDoc)
```
Comment (jolshan) [inline core/src/main/scala/kafka/server/KafkaConfig.scala:1328]: Do we want to enforce that the min value is less than or equal to the max value?

File: core/src/test/scala/unit/kafka/server/KafkaConfigTest.scala [hunk 2]
```diff
@@ -976,6 +976,20 @@ class KafkaConfigTest {
         case RemoteLogManagerConfig.REMOTE_LOG_READER_THREADS_PROP => assertPropertyInvalid(baseProperties, name, "not_a_number", 0, -1)
         case RemoteLogManagerConfig.REMOTE_LOG_READER_MAX_PENDING_TASKS_PROP => assertPropertyInvalid(baseProperties, name, "not_a_number", 0, -1)
 
+        /** New group coordinator configs */
+        case KafkaConfig.NewGroupCoordinatorEnableProp => // ignore string
+        case KafkaConfig.GroupCoordinatorNumThreadsProp => assertPropertyInvalid(baseProperties, name, "not_a_number", 0, -1)
+
+        /** Consumer groups configs */
+        case KafkaConfig.ConsumerGroupSessionTimeoutMsProp => // ignore string
```
Comment (jolshan) [inline core/src/test/scala/unit/kafka/server/KafkaConfigTest.scala]: Should this one assert the same as the ones below.

File: core/src/test/scala/unit/kafka/server/KafkaConfigTest.scala [hunk 3]
```diff
@@ -976,6 +976,20 @@ class KafkaConfigTest {
         case RemoteLogManagerConfig.REMOTE_LOG_READER_THREADS_PROP => assertPropertyInvalid(baseProperties, name, "not_a_number", 0, -1)
         case RemoteLogManagerConfig.REMOTE_LOG_READER_MAX_PENDING_TASKS_PROP => assertPropertyInvalid(baseProperties, name, "not_a_number", 0, -1)
 
+        /** New group coordinator configs */
+        case KafkaConfig.NewGroupCoordinatorEnableProp => // ignore string
```
Comment (jolshan) [inline core/src/test/scala/unit/kafka/server/KafkaConfigTest.scala]: nit: this one isn't a string -- some of the other configs just said "ignore"

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 4]
```diff
@@ -1267,6 +1314,24 @@ object KafkaConfig {
       .define(GroupInitialRebalanceDelayMsProp, INT, Defaults.GroupInitialRebalanceDelayMs, MEDIUM, GroupInitialRebalanceDelayMsDoc)
       .define(GroupMaxSizeProp, INT, Defaults.GroupMaxSize, atLeast(1), MEDIUM, GroupMaxSizeDoc)
 
+      /** New group coordinator configs */
+      // All properties are kept internal until KIP-848 is released.
+      // This property meant to be here only during the development of KIP-848. It will
+      // be replaced by a metadata version before releasing it.
+      .defineInternal(NewGroupCoordinatorEnableProp, BOOLEAN, Defaults.NewGroupCoordinatorEnable, null, MEDIUM, NewGroupCoordinatorEnableDoc)
```
Comment (jolshan) [inline core/src/main/scala/kafka/server/KafkaConfig.scala:1321]: nit: you can leave "null" out (same with the other prop below)

**Comment by jolshan** [comment]
Just a general question about configs -- do we have any guidelines for adding and removing configs between releases? Just curious about compatibility and typical guidelines around that.

**Comment by dajac** [comment]
> Just a general question about configs -- do we have any guidelines for adding and removing configs between releases? Just curious about compatibility and typical guidelines around that.

Configs are treated as public APIs so we treat them as such. This is why I keep all of them internal for now. They won't be published until we remove this.

**Comment by jolshan** [comment]
Got it. So the distinction is "publishing" and that is done when the config is not internal. Thanks for clarifying.

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 5]
```diff
@@ -163,6 +163,20 @@ object Defaults {
...
+  val NewGroupCoordinatorEnable = false
+  val GroupCoordinatorNumThreads = 1
+
+  /** Consumer group configs */
+  val ConsumerGroupSessionTimeoutMs = 45000
+  val ConsumerGroupMinSessionTimeoutMs = 45000
+  val ConsumerGroupMaxSessionTimeoutMs = 60000
+  val ConsumerGroupHeartbeatIntervalMs = 5000
+  val ConsumerGroupMinHeartbeatInternalMs = 5000
+  val ConsumerGroupMaxHeartbeatInternalMs = 15000
```
Comment (jeffkbkim) [inline core/src/main/scala/kafka/server/KafkaConfig.scala]: should these be "Interval"?

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 6]
```diff
@@ -163,6 +163,20 @@ object Defaults {
...
+  val ConsumerGroupMaxSize = Int.MaxValue
+  val ConsumerGroupAssignors = ""
```
Comment (jeffkbkim) [inline core/src/main/scala/kafka/server/KafkaConfig.scala:178]: KIP 848 lists
```
org.apache.kafka.server.group.consumer.UniformAssignor, org.apache.kafka.server.group.consumer.RangeAssignor
```
as the default. are we waiting until these two are implemented?

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 7]
```diff
@@ -483,11 +497,27 @@ object KafkaConfig {
...
+  val NewGroupCoordinatorEnableProp = "group.coordinator.new.enable"
+  val GroupCoordinatorNumThreadsProp = "group.coordinator.threads"
+
+  /** Consumer group configs */
+  val ConsumerGroupSessionTimeoutMsProp = "group.consumer.session.timeout.ms"
+  val ConsumerGroupMinSessionTimeoutMsProp = "group.consumer.min.session.timeout.ms"
+  val ConsumerGroupMaxSessionTimeoutMsProp = "group.consumer.max.session.timeout.ms"
+  val ConsumerGroupHeartbeatIntervalMsProp = "group.consumer.heartbeat.interval.ms"
+  val ConsumerGroupMinHeartbeatInternalMsProp = "group.consumer.min.heartbeat.interval.ms"
+  val ConsumerGroupMaxHeartbeatInternalMsProp ="group.consumer.max.heartbeat.interval.ms"
```
Comment (jeffkbkim) [inline core/src/main/scala/kafka/server/KafkaConfig.scala]: should these also be "Interval"?

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 8]
```diff
@@ -1267,6 +1314,24 @@ object KafkaConfig {
       .define(GroupInitialRebalanceDelayMsProp, INT, Defaults.GroupInitialRebalanceDelayMs, MEDIUM, GroupInitialRebalanceDelayMsDoc)
       .define(GroupMaxSizeProp, INT, Defaults.GroupMaxSize, atLeast(1), MEDIUM, GroupMaxSizeDoc)
 
+      /** New group coordinator configs */
+      // All properties are kept internal until KIP-848 is released.
+      // This property meant to be here only during the development of KIP-848. It will
```
Comment (jeffkbkim) [inline core/src/main/scala/kafka/server/KafkaConfig.scala]: nit: "property is meant"

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 9]
```diff
@@ -1267,6 +1314,24 @@ object KafkaConfig {
...
+      // be replaced by a metadata version before releasing it.
+      .defineInternal(NewGroupCoordinatorEnableProp, BOOLEAN, Defaults.NewGroupCoordinatorEnable, null, MEDIUM, NewGroupCoordinatorEnableDoc)
+      .defineInternal(GroupCoordinatorNumThreadsProp, INT, Defaults.GroupCoordinatorNumThreads, atLeast(1), MEDIUM, GroupCoordinatorNumThreadsDoc)
+
+      /** Consumer groups configs */
+      // All properties are kept internal until KIP-848 is released.
+      .defineInternal(ConsumerGroupSessionTimeoutMsProp, INT, Defaults.ConsumerGroupSessionTimeoutMs, atLeast(1), MEDIUM, ConsumerGroupSessionTimeoutMsDoc)
```
Comment (jeffkbkim) [inline core/src/main/scala/kafka/server/KafkaConfig.scala:1326]: how does the operator use ConsumerGroupSessionTimeoutMs and ConsumerGroupSessionMin/MaxTimeoutMs configs?

is the idea to set min/max once, then configure the base timeout which the min/max would enforce?

**Review by jeffkbkim — COMMENTED** [review]
thanks, left some comments

File: core/src/main/scala/kafka/server/KafkaConfig.scala (hunk 6 above)
Comment (dajac) [reply core/src/main/scala/kafka/server/KafkaConfig.scala:178]: Correct. We will add them when we have them.

File: core/src/main/scala/kafka/server/KafkaConfig.scala (hunk 4 above)
Comment (dajac) [reply core/src/main/scala/kafka/server/KafkaConfig.scala:1321]: `defineInternal` does not have an overload without the `validator`.

File: core/src/main/scala/kafka/server/KafkaConfig.scala (hunk 9 above)
Comment (dajac) [reply core/src/main/scala/kafka/server/KafkaConfig.scala:1326]: They will be mainly used to ensure that the dynamic group configurations are within the defined boundaries.

**Comment by dajac** [comment]
@jolshan @jeffkbkim Thanks for your comments. Addressed them.

File: core/src/main/scala/kafka/server/KafkaConfig.scala (hunk 4 above)
Comment (jolshan) [reply core/src/main/scala/kafka/server/KafkaConfig.scala:1321]: ah interesting. Thanks for clarifying.

**Review by jolshan — APPROVED** [review]
thanks!

**Pull request merged**